Order Status Mark 1: Add new entity and change attribute name

mindgraffiti opened the pull request:
Changed the OrderStatus enum to OrderStatusKey. Added the new entity OrderStatus (but it's not used or mapped anywhere yet, in Networking). This PR is mostly generated code.

## To Test
1. Delete app, reset Content & Settings
2. Check out PR and build
3. Run unit tests and ensure they're green

cc: @astralbodies and @bummytime ready for review!

Update release notes:
- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

bummytime commented:
@mindgraffiti another quick thing that stuck out to me was this:

<img width="317" alt="screen shot 2019-02-21 at 7 32 00 pm" src="https://user-images.githubusercontent.com/154014/53213770-d061f200-360f-11e9-9d37-5f2d05c499f9.png">

I see a 1-to-1 relationship here — at a high level it is usually best to avoid these when possible (in Core Data and in DB models in general 😄 ). I have not researched this as much as you, but perhaps there is a way to either have a lookup table of statuses with no relationship to order? Perhaps something else?

astralbodies commented:
> I see a 1-to-1 relationship here — at a high level it is usually best to avoid these when possible (in Core Data and in DB models in general 😄 ). I have not researched this as much as you, but perhaps there is a way to either have a lookup table of statuses with no relationship to order? Perhaps something else?

If `OrderStatus` is to be shared between multiple `Order`s then many to one makes sense. I'd even say turn the `SiteID` property into a relationship with `Site` if that doesn't break any previous decisions with relationships in the Storage layer.

mindgraffiti commented:
@astralbodies the errors that @bummytime is seeing are the same ones that were making my unit tests fail yesterday. I'll ping you later today

bummytime commented:
@astralbodies @mindgraffiti Just for a point of reference...the site settings is stand-alone entity — no relationship back to site.  

<img width="329" alt="screen shot 2019-02-22 at 9 13 37 am" src="https://user-images.githubusercontent.com/154014/53251617-b1507800-3682-11e9-8dec-093a21e79e65.png">

Overall, Jorge and I tried to reduce the number of Core Data relationships because having a spider web of relationships in CD can be a nightmare to manage and also is a vector for additional bugs.

mindgraffiti commented:
>stand-alone entity

Ok, so the Order.statusKey is supposed to function as the slug / key for looking up OrderStatuses. I didn't realize they could be standalone entities. That's good news!

bummytime commented:
> Ok, so the Order.statusKey is supposed to function as the slug / key for looking up OrderStatuses. I didn't realize they could be standalone entities. That's good news!

@mindgraffiti that may mean you have to do a little more legwork in the related Yosemite stores, but yeah, anything is possible with Yosemite. Right @jleandroperez?? :trollface: 😆

mindgraffiti commented:
Ok, added missing source for statusKey, made OrderStatus a standalone entity and deleted the siteID attribute from OrderStatus. @bummytime @astralbodies anything else required on this before a second review?

bummytime commented on WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift:
```diff
@@ -150,7 +150,7 @@ private extension OrdersViewController {
         resultsController.predicate = {
             let excludeSearchCache = NSPredicate(format: "exclusiveForSearch = false")
-            let excludeNonMatchingStatus = statusFilter.map { NSPredicate(format: "status = %@", $0.rawValue) }
+            let excludeNonMatchingStatus = statusKeyFilter.map { NSPredicate(format: "status = %@", $0.rawValue) }
```
```suggestion
            let excludeNonMatchingStatus = statusKeyFilter.map { NSPredicate(format: "statusKey = %@", $0.rawValue) }
```

@mindgraffiti we need to change the predicate format string to `statusKey` otherwise filtering causes the app to crash. 💣

mindgraffiti commented:
Thank you so much @bummytime and @astralbodies!
